### PR TITLE
Create new permission api with validations

### PR DESCRIPTION
**Pull request by subash-cs**

### CODE REVIEW

File: SanitySuiteSCP_API.xml
Comment (infa-rvarahag) [inline SanitySuiteSCP_API.xml:1]: class name should be  : TestSCPermissionCRUPOperations
xml syntax should be like this :
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="SanitySuite">

    <test name="Test">
        <classes>

            <class name="org.shikshalokam.backend.ep.CLASSNAME">
                <methods>
                    <include name="TestMethodNames1"/>
                    <include name="TestMethodNames1"/>
                    <include name="TestMethodNames1"/>
                </methods>
            </class>

        </classes>
    </test> 
</suite>

File: src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java [hunk 1]
```diff
@@ -33,11 +33,10 @@ public void init() {
...
-        logger.info(response.prettyPrint());
     }
 
-    @Test(description = "Verifies the functionality of creating new user's permission.")
-    public void testCreateNewPermission() {
+    @Test(description = "Verifies the functionality of creating new user's permission with valid payload.")
+    public void testCreatePermissionWithValidPayload() {
         logger.info("Started calling the CreatePermission API:");
         try {
             RestAssured.baseURI = PropertyLoader.PROP_LIST.get("scp.qa.api.base.url").toString();
```
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java]: This already we are doing in the base call hence it is not required again here . Remove this

File: src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java [hunk 2]
```diff
@@ -48,7 +47,7 @@ public void testCreateNewPermission() {
 
         // Updated request body
         String requestBody = "{ " +
-                "\"code\": \"project_subash_uat\", " +
+                "\"code\": \"project_test_break\", " +
```
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java]: Please use code like this to create a Json object and pass the same as body to the request 

JSONObject jsonObject = new JSONObject();

        // Add key-value pairs
        jsonObject.put("name", "John Doe");
        jsonObject.put("age", 30);
        jsonObject.put("email", "johndoe@example.com");
        jsonObject.put("isMarried", false);

File: src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java [hunk 3]
```diff
@@ -62,10 +61,95 @@ public void testCreateNewPermission() {
...
+        // Log the status code and response body
+        int statusCode = response.getStatusCode();
+        String responseBody = response.getBody().asString();
+        logger.info("Response Status Code: " + statusCode);
+        logger.info("Response Body: " + responseBody);
+        // Pretty print the response to the console or logs
+        response.prettyPrint();
+
+        //log the end call for createPermission API
+        logger.info("Ended calling the CreatePermission API with valid payload.");
```
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java]: Please add 
assertEquals to validate response code .
fetch responce and validate ID 
No assertions

File: src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java [hunk 4]
```diff
@@ -62,10 +61,95 @@ public void testCreateNewPermission() {
...
+        logger.info("Started calling the CreatePermission API:");
+        try {
+            RestAssured.baseURI = PropertyLoader.PROP_LIST.get("scp.qa.api.base.url").toString();
+            createPermissionEndpoint = new URI(PROP_LIST.get("scp.create.permission.endpoint").toString());
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+
+        // Updated request body
+        String requestBody = "{ " +
```
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java]: Remove String and add Json Object structure

File: src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java [hunk 5]
```diff
@@ -62,10 +61,95 @@ public void testCreateNewPermission() {
...
+        //log the end call for createPermission API
+        logger.info("Ended calling the CreatePermission API with valid payload.");
+
+    }
+
+    @Test(description = "Verifies the functionality of permission creation when the payload contains invalid payload")
+    public void testCreatePermissionWithInvalidPayload() {
+        logger.info("Started calling the CreatePermission API:");
+        try {
+            RestAssured.baseURI = PropertyLoader.PROP_LIST.get("scp.qa.api.base.url").toString();
```
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java]: Remove this , this is not required , while login base URL already assigned by extended class

File: src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java [hunk 6]
```diff
@@ -62,10 +61,95 @@ public void testCreateNewPermission() {
...
+        int statusCode = response.getStatusCode();
+        String responseBody = response.getBody().asString();
+        logger.info("Response Status Code: " + statusCode);
+        logger.info("Response Body: " + responseBody);
+        // Pretty print the response to the console or logs
+        response.prettyPrint();
+
+        //log the end call for createPermission API
+        logger.info("Ended calling the CreatePermission API with invalid payload.");
+
```
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java]: Please validate the response messages . code using TestNG assertions

File: src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java [hunk 7]
```diff
@@ -62,10 +61,95 @@ public void testCreateNewPermission() {
...
+    }
+
+    @Test(description = "Verifies the functionality of permission creation when the payload contains empty fields")
+    public void testCreatePermissionWithEmptyFields() {
+        logger.info("Started calling the CreatePermission API:");
+        try {
+            RestAssured.baseURI = PropertyLoader.PROP_LIST.get("scp.qa.api.base.url").toString();
```
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java]: Remove this not required

File: src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java [hunk 8]
```diff
@@ -62,10 +61,95 @@ public void testCreateNewPermission() {
...
+            createPermissionEndpoint = new URI(PROP_LIST.get("scp.create.permission.endpoint").toString());
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+
+        // Updated request body
+        String requestBody = "{ " +
```
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java]: Convert to JsonObject , remove string

File: src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java [hunk 9]
```diff
@@ -62,10 +61,95 @@ public void testCreateNewPermission() {
...
+        // Log the status code and response body
+        int statusCode = response.getStatusCode();
+        String responseBody = response.getBody().asString();
+        logger.info("Response Status Code: " + statusCode);
+        logger.info("Response Body: " + responseBody);
+        // Pretty print the response to the console or logs
+        response.prettyPrint();
+
+        //log the end call for createPermission API
+        logger.info("Ended calling the CreatePermission API with empty fields payload.");
```
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/SelfCreationPortalCreateNewPermission.java]: Please add assertions to the response using testNG

**Review by infa-rvarahag — CHANGES_REQUESTED** [review]

File: src/test/java/org/shikshalokam/backend/scp/TestScpPermissionCRUDOperations.java
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/TestScpPermissionCRUDOperations.java:1]: Move the Base url assignment to  Base class 
Remove the code for fetching internal access token and start using the base class assigned token X_AUTH_TOKEN

File: src/test/java/org/shikshalokam/backend/scp/TestScpPermissionCRUDOperations.java [hunk 10]
```diff
@@ -39,117 +39,139 @@ public void init() {
...
+        // Fetch ID from response and validate
+        int createdId = response.jsonPath().getInt("result.id");
+        Assert.assertNotNull(createdId, "Created ID should not be null");
+        logger.info("Created ID: " + createdId);
+
+        // Log the end call for createPermission API
         logger.info("Ended calling the CreatePermission API with valid payload.");
+    }
 
+    //Method to generate random word as code(Project name)
```
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/TestScpPermissionCRUDOperations.java]: RandomStringUtils.randomAlphabetic(20) use this method .Bulit in methods more optimized

File: src/test/java/org/shikshalokam/backend/scp/TestScpPermissionCRUDOperations.java
Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/TestScpPermissionCRUDOperations.java:1]: write a common method  for the Respsonce createpermission ( JsonObject  object )

then start using this in all Test methods and validate the response written

Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/TestScpPermissionCRUDOperations.java:1]: Move all non test methods to the end of the class

Comment (infa-rvarahag) [inline src/test/java/org/shikshalokam/backend/scp/TestScpPermissionCRUDOperations.java:1]: move to Backend

**Review by infa-rvarahag — CHANGES_REQUESTED** [review]

**Review by subash-cs — COMMENTED** [review]
Comments done

**Review by infa-rvarahag — APPROVED** [review]

**Pull request merged**